Rebase LSTOD

ariostas opened the pull request:
This is a rebase of https://github.com/SegmentLinking/TrackLooper/pull/412.

It's a very large diff, so @Hoobidoobidoo will have to make sure that it's right. It should be a good starting point to finish the work, so I'll let @Hoobidoobidoo take over. I'm happy to answer questions or help with something if needed.

slava77 commented:
> This branch has conflicts that must be resolved

apparently after recent LST PRs

Hoobidoobidoo commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1172,9 +1223,10 @@ std::tuple<float, float, float, std::vector<unsigned int>, std::vector<unsigned
-  float pt = segmentsPixel.ptIn()[pLS];
-  float eta = segmentsPixel.eta()[pLS];
-  float phi = segmentsPixel.phi()[pLS];
+  const int seedIdx = segmentsPixel.seedIdx()[pLS];
```
what was the motivation for switching the indexing from pLS to seedIdx in these cases?

ariostas replied:
That was a change that @sgnoohc made in [this commit](https://github.com/Hoobidoobidoo/TrackLooper/commit/13b0fd1c99b662e894ff63e31d9e781bd885018e). I'm not sure what was the reasoning.

Hoobidoobidoo replied:
I see, Philip was not really sure either, but he said they are probably the same thing. Thank you for clarifying.

ariostas replied:
Okay, I'll revert this change to keep things more consistent.

Hoobidoobidoo commented:
I think everything is good on my end to push this pr.

ariostas commented:
@Hoobidoobidoo instead of adding additional options and flags, I decided to use the existing `CUT_VALUE_DEBUG` flag. So to enable the extra outputs you can just use the `-d` flag. Let me know if this would be an issue for you.

I verified that it compiles (after fixing a couple of typos I made), but could you also verify that the output ntuple contains all the data you want and that it looks good?

ariostas commented:
/run all

slava77 commented:
was there a feature change in this PR in the algorithm ( I thought this was just ntuple-related)
why is the efficiency changing?
https://raw.githubusercontent.com/SegmentLinking/TrackLooper-plots-archive/cmssw/PR149_43311a5_standalone/eff_pt_comp.png

ariostas commented:
> was there a feature change in this PR in the algorithm ( I thought this was just ntuple-related)
> why is the efficiency changing?

No, there was no change in the algorithm. There must be a bug somewhere

Hoobidoobidoo commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1351,3 +1400,1559 @@ void printT3s(LSTEvent* event) {
+    float bestmatch_frac = -999;
+    for (size_t jj = 0; jj < sim_tcIdxAll_to_write.at(i).size(); ++jj) {
+      int idx = sim_tcIdxAll_to_write.at(i).at(jj);
+      float frac = sim_tcIdxAll_to_write.at(i).at(jj);
```
Should be sim_tcIdxAllFrac

ariostas replied:
Ah okay, thanks for finding this!

I'll mark the PR as ready for review to see if there are any conflicts, and I'll fix this.

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1351,3 +1400,1559 @@ void printT3s(LSTEvent* event) {
+    float bestmatch_frac = -999;
+    for (size_t jj = 0; jj < sim_tcIdxAll_to_write.at(i).size(); ++jj) {
+      int idx = sim_tcIdxAll_to_write.at(i).at(jj);
+      float frac = sim_tcIdxAll_to_write.at(i).at(jj);
```
```suggestion
      float frac = sim_tcIdxAllFrac_to_write.at(i).at(jj);
```

ariostas commented:
/run all

ariostas commented:
/run all

ariostas commented on RecoTracker/LSTCore/standalone/efficiency/src/performance.cc:
```diff
@@ -675,6 +740,7 @@ void fillFakeRateSets(std::vector<RecoTrackSetDefinition>& FRsets) {
 
 //__________________________________________________________________________________________________________________________________________________________________________
 void fillFakeRateSet(int itc, RecoTrackSetDefinition& FRset) {
+  std::cout << "pt variable size" << FRset.pt().size() << std::endl;
```
We need to remove this line

ariostas commented:
The cpu timings on cgpu-1 also look good.

```
This PR (357dc9f)
Total Timing Summary
Average time for map loading = 349.674 ms
Average time for input loading = 8143.17 ms
Average time for lst::Event creation = 0.00399635 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     19.8    281.5    141.6    143.3     33.5    363.8     77.8     93.2    107.7      1.1    1263.3     879.7+/- 220.8    1264.3   explicit[s=1]
   avg     19.3    280.3    139.6    146.2     32.8    359.7     76.4     92.6    106.9      0.9    1254.8     875.7+/- 212.7     329.5   explicit[s=4]
   avg     20.8    291.4    145.0    147.8     34.9    370.2     80.0     95.7    109.6      1.1    1296.3     905.3+/- 217.0      94.9   explicit[s=16]
   avg     22.6    295.4    146.6    149.3     35.8    374.6     80.6     97.7    111.2      1.0    1314.9     917.7+/- 220.4      53.8   explicit[s=32]
   avg     32.3    312.6    151.6    158.8     39.1    398.0     84.0    101.8    115.8      1.2    1395.2     964.9+/- 228.6      32.7   explicit[s=64]

master (538ac87)
Total Timing Summary
Average time for map loading = 341.615 ms
Average time for input loading = 200098 ms
Average time for lst::Event creation = 0.00371252 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     18.4    261.5    133.1    132.6     30.5    333.6     70.6     85.6     98.6      0.8    1165.3     813.3+/- 194.7    1166.5   explicit[s=1]
   avg     18.1    260.5    131.9    130.6     30.8    332.8     72.4     85.7     99.1      0.8    1162.8     811.9+/- 194.4     296.0   explicit[s=4]
   avg     20.5    288.7    143.6    144.8     34.4    366.7     79.3     94.7    108.6      1.0    1282.2     895.0+/- 214.1      93.9   explicit[s=16]
   avg     22.6    295.2    147.8    150.2     35.9    374.4     80.7     97.7    111.3      1.0    1316.8     919.8+/- 217.2      53.5   explicit[s=32]
   avg     32.4    312.6    152.0    159.1     39.3    405.7     83.9    102.2    119.3      1.1    1407.5     969.5+/- 230.3      32.8   explicit[s=64]
```

slava77 commented:
> The cpu timings on cgpu-1 also look good.

was this tested with `CUT_VALUE_DEBUG` enabled?

slava77 commented:
> was this tested with `CUT_VALUE_DEBUG` enabled?

a related question: is there a way to ask for it from the CI /run ...?

slava77 commented on RecoTracker/LSTCore/standalone/efficiency/src/LSTEff.cc:
```diff
@@ -3,46 +3,39 @@ LSTEff lstEff;
-    t3_phi_branch = tree->GetBranch("t3_phi");
-    if (t3_phi_branch) {
-      t3_phi_branch->SetAddress(&t3_phi_);
+  sim_pt_branch = 0;
+  if (tree->GetBranch("sim_pt") != 0) {
```
is this auto-generated or manually ported/merged?

ariostas replied:
This was manually ported. I didn't realize that this file was auto-generated. I feel like this should be refactored to make it a lot more concise and manageable.

slava77 commented on RecoTracker/LSTCore/standalone/eventdisplay/environment.yml:
this is the only file in this directory.
- Is it up to date?
- how to generate a new one if some dependencies change?

If all in the dependencies is auto-generated, it may be better to generate the file as well (and have some instructions in a readme)

ariostas replied:
I don't think it makes sense to include this file here. Judging by the name, this is for the eventdisplay package, so it would make more sense there.

slava77 commented on RecoTracker/LSTCore/standalone/efficiency/src/LSTEff.h:
do we have already some documentation how to remake this file?
Considering some lines are showing up with `CUTVALUE_DEBUG`, it looks like some manual editing was in place.
Maybe longer term it's better to bootstrap the autogenerated parts as a part of the area setup/install.

ariostas replied:
This is fixed now

slava77 commented:
my review of the /standalone changes was rather cursory.
Perhaps a second look from others would help as well.

It would be good to understand what the impact is for the standard benchmark outputs.
Do I recall correctly that the timing benchmark is done without writing the ntuple?

How much slower is the case with an ntuple when default the validation plots are made (and what is the file size increase in this case)? Same for the case with `CUTVALUE_DEBUG` enabled.

ariostas commented:
Thanks for all the comments, Slava!

> was this tested with CUT_VALUE_DEBUG enabled?

No, but I'll leave timings with it enabled below.

> a related question: is there a way to ask for it from the CI /run ...?

No, but if it becomes a common need I can add that. In that case we should also add the option to `lst_timing`.

> Perhaps a second look from others would help as well.

I'll also take a look. While rebasing I fixed some obvious things, but I didn't look at it very carefully.

> It would be good to understand what the impact is for the standard benchmark outputs.
> Do I recall correctly that the timing benchmark is done without writing the ntuple?

That's right. I'll also run the timing comparison with writing enabled.



I'll address some of the comments you left, but I'll refer the rest to @Hoobidoobidoo or @sgnoohc since they can better address them.

Heres a timing comparison on GPU with `CUTVALUE_DEBUG` enabled.
```
This PR (357dc9f)
Total Timing Summary
Average time for map loading = 548.513 ms
Average time for input loading = 7668.05 ms
Average time for lst::Event creation = 0.00314485 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      1.3      0.5      0.4      1.8      0.6      0.5      0.9      0.5      1.1      0.0       7.6       5.8+/-  1.1       9.6   explicit_cutvalue[s=1]
   avg      1.3      0.7      0.6      2.2      0.8      0.7      1.2      0.7      1.5      0.0       9.6       7.6+/-  1.6       5.9   explicit_cutvalue[s=2]
   avg      2.3      1.1      1.1      3.1      1.6      1.1      2.2      1.1      2.5      0.0      16.0      12.7+/-  2.8       4.6   explicit_cutvalue[s=4]
   avg      3.7      1.8      1.7      4.2      2.6      1.6      3.1      1.5      3.8      0.0      23.8      18.6+/-  4.0       4.5   explicit_cutvalue[s=6]
   avg      5.5      2.7      2.4      5.2      3.6      2.0      4.4      2.0      4.9      0.0      32.7      25.3+/-  7.0       4.5   explicit_cutvalue[s=8]

master (538ac87)
Total Timing Summary
Average time for map loading = 547.471 ms
Average time for input loading = 7653.65 ms
Average time for lst::Event creation = 0.00329245 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      1.3      0.5      0.4      1.8      0.6      0.5      0.9      0.5      1.1      0.0       7.6       5.8+/-  1.1       9.6   explicit_cutvalue[s=1]
   avg      1.3      0.7      0.6      2.1      0.8      0.7      1.1      0.7      1.5      0.0       9.5       7.5+/-  1.6       5.9   explicit_cutvalue[s=2]
   avg      2.2      1.2      1.1      3.1      1.6      1.1      2.2      1.1      2.5      0.0      16.1      12.8+/-  2.8       4.7   explicit_cutvalue[s=4]
   avg      3.5      1.9      1.7      4.1      2.5      1.6      3.2      1.4      3.7      0.0      23.7      18.6+/-  4.7       4.5   explicit_cutvalue[s=6]
   avg      5.5      2.8      2.3      5.1      3.6      2.1      3.9      2.0      4.9      0.0      32.1      24.5+/-  5.7       4.4   explicit_cutvalue[s=8]
```

ariostas commented:
Here is a timing comparison with `CUTVALUE_DEBUG` and writing enabled.

```
This PR (357dc9f)
Total Timing Summary
Average time for map loading = 590.796 ms
Average time for input loading = 7655.72 ms
Average time for lst::Event creation = 0.00303133 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      1.6      0.5      0.4      1.8      0.7      0.5      0.9      0.5      1.1      0.1       8.1       6.0+/-  1.2    2673.2   explicit_cutvalue[s=1]
   avg      1.4      0.5      0.4      1.8      0.7      0.5      1.0      0.6      1.1      0.1       8.0       6.1+/-  1.4    2857.4   explicit_cutvalue[s=2]
   avg      1.8      0.5      0.5      1.9      0.7      0.6      1.1      0.6      1.1      0.1       8.9       6.5+/-  3.2    2837.8   explicit_cutvalue[s=4]
   avg      2.5      0.6      0.6      2.0      0.8      0.6      1.3      0.8      1.3      0.1      10.5       7.4+/-  5.8    3140.6   explicit_cutvalue[s=6]
   avg      2.7      0.7      0.5      2.1      0.9      0.6      1.8      0.8      1.3      0.1      11.5       8.2+/-  9.6    2724.8   explicit_cutvalue[s=8]

master (538ac87)
Total Timing Summary
Average time for map loading = 568.772 ms
Average time for input loading = 8300.4 ms
Average time for lst::Event creation = 0.00299726 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg      3.1      0.6      0.6      2.0      0.8      0.6      1.5      1.1      1.2      0.1      11.7       8.0+/-  1.5    2859.7   explicit_cutvalue[s=1]
   avg      2.0      0.6      0.6      2.1      0.9      0.6      1.4      0.9      2.2      0.1      11.5       8.8+/-  3.5    2592.4   explicit_cutvalue[s=2]
   avg      4.5      0.8      0.8      2.3      1.1      0.8      2.4      2.4      3.8      0.2      19.1      13.8+/-  6.7    2875.7   explicit_cutvalue[s=4]
   avg      3.3      0.8      0.7      2.2      0.9      0.7      2.3      1.6      2.2      0.1      14.6      10.7+/-  7.3    2565.5   explicit_cutvalue[s=6]
   avg      4.0      0.9      0.7      2.4      1.1      0.8      2.5      1.6      2.2      0.1      16.0      11.3+/- 10.4    2604.3   explicit_cutvalue[s=8]
```

ariostas commented on RecoTracker/LSTCore/src/alpaka/Segment.h:
```diff
@@ -493,7 +534,22 @@ namespace ALPAKA_ACCELERATOR_NAMESPACE::lst {
+                                                            float& rtHi,
+#endif
                                                             const float ptCut) {
+#ifdef CUT_VALUE_DEBUG
+    zLo = -999.f;
+    zHi = -999.f;
+    rtLo = -999.f;
+    rtHi = -999.f;
+#endif
```
Yeah, that's a good idea

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.h:
```diff
@@ -59,4 +61,8 @@ void printT3s(LSTEvent* event);
 void printTCs(LSTEvent* event);
 
+// LSTOD
+void createOutputBranches_v2();
```
Do we want to keep the old version too or just replace the function?

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -21,7 +25,7 @@ void fillOutputBranches(LSTEvent* event) {
   ana.tx->fill();
 
-  // Then clear the branches to default values (e.g. -999, or clear the vectors to empty vectors)
+  // Then clear the branches to default values (e.g. -999, or clear the std::vectors to empty std::vectors)
```
Yeah, that was just me find-and-replacing all instances. I fixed all the comments now.

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1486,3 +1535,1559 @@ void printT3s(LSTEvent* event) {
+      float dphichange = miniDoublets.dphichanges()[mdIdx];
+      float dphi = miniDoublets.dphis()[mdIdx];
+      float dz = miniDoublets.dzs()[mdIdx];
+      float k2Rinv1GeVf = (2.99792458e-3 * 3.8) / 2;
```
This one is defined in the `alpaka/Common.h`. Maybe we should think how to share constants between host and device code

ariostas commented:
I fixed a few issues with the script that generates `LSTEff.h` and `LSTEff.cc` and there were a few typos in those files. I'll push these updates once @Hoobidoobidoo pushes his (which he should be able to do now).

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1486,3 +1535,1559 @@ void printT3s(LSTEvent* event) {
+      // Compute whether this is a fake
+      bool isfake = true;
+      for (size_t isim = 0; isim < simidx.size(); ++isim) {
+        if (simidxfrac[isim] > 0.75) {
```
Hubert took care of this

Hoobidoobidoo commented:
/run all

Hoobidoobidoo commented:
/run all

Hoobidoobidoo commented:
I think everything looks good to the output given the changes. Let me know if we are good to push it through.

GNiendorf commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -1486,3 +1534,1532 @@ void printT3s(LSTEvent* event) {
+  //--------------------------------------------
+  //
+  //
+  // Mini-Doublets
```
This is my first time looking at this PR, but is there any way these can be moved to separate functions for the different objects? Seems like a organizational downgrade to move the code to a single function call for all objects.

slava77 commented:
> I see, I can remove the v1 code, I will also check on the needed branches for DNN

please clarify on the status from your side.

sgnoohc commented:
So far I think the two issues are:
- output file size are large => does it need to be default?
- the function is larger => can we modularize it.

Aside from this, are there any performance / validation issue?

ariostas commented:
This is pretty much done now. I'll just do some minor cleanup. @sgnoohc @Hoobidoobidoo could you check that the output ntuple looks good?

/run gpu-all
/run checks

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -319,20 +560,167 @@ void setOutputBranches(LSTEvent* event) {
+      }
+    }
+
+    // TODO: should this be enabled or removed?
```
@sgnoohc should this section be enabled or just removed?

And more generally, is it fine to remove lines that are commented out?

slava77 commented:
/run all

slava77 reviewed (approved):
I don't have substantive comments after a somewhat cursory look.

I mentioned during the meeting to check the time to process of the (default) validation workflow; but after re-checking the actions I see that we run lst only once for both validation ntuples and for timing in the CPU standalone workflow `lst_cpu -i PU200 -o LSTNtuple_after.root -s 4 -v 1 $LOW_PT_FLAG`. While the actual writing is not counted, the branch filling is collected by the [full_timer](https://github.com/cms-sw/cmssw/blob/6499255bddf686deda739048f4a201ce38eb6395/RecoTracker/LSTCore/standalone/bin/lst.cc#L477C9-L477C19) and will show up in the `Rate` column.
There is about 6% increase in the [last cpu test](https://github.com/SegmentLinking/cmssw/pull/149#issuecomment-3025166953) or 34 seconds, which seems acceptable.

GNiendorf commented:
Could we double check that the NN notebooks train fine still after this PR? Also, does the timing with -w 1 (the ntuple write-out) increase substantially? Those are my two general concerns for this PR.

ariostas commented:
The "standard" timing doesn't change significantly because it writes about the same data as before (although it increases enough to be a bottleneck for multiple streams). To enable the extra output you compile it with `-d` and it that case it does slow down significantly. @GNiendorf would you be able to check the notebooks since you're more familiar with them?

Here are timing tests on cgpu-1.

This is with `-w 1`.
```
master
Total Timing Summary
Average time for map loading = 594.484 ms
Average time for input loading = 16018.3 ms
Average time for lst::Event creation = 0.000715256 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     14.0      0.5      0.4      0.7      0.8      0.7      0.8      0.4      1.1      0.0      19.5       4.9+/-  0.9     207.1   explicit[s=1]
   avg      1.3      0.5      0.4      0.7      0.8      0.7      0.9      0.5      1.1      0.0       7.0       5.0+/-  1.4     141.6   explicit[s=2]
   avg      1.6      0.5      0.5      0.8      0.9      0.7      1.1      0.5      1.1      0.0       7.8       5.4+/-  3.4     139.1   explicit[s=4]
   avg      2.3      0.6      0.5      0.9      1.0      0.7      1.4      0.6      1.2      0.0       9.1       6.1+/-  6.2     138.8   explicit[s=6]
   avg      2.7      0.7      0.5      0.9      1.0      0.7      1.4      0.8      1.3      0.0      10.0       6.6+/-  7.7     141.1   explicit[s=8]


This PR
Total Timing Summary
Average time for map loading = 592.162 ms
Average time for input loading = 7877.47 ms
Average time for lst::Event creation = 0.000669843 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     14.8      0.5      0.4      0.7      0.8      0.6      0.9      0.4      1.1      0.0      20.3       4.9+/-  0.9     206.5   explicit[s=1]
   avg      1.2      0.5      0.4      0.8      0.8      0.7      0.9      0.5      1.1      0.0       6.9       5.0+/-  1.3     196.4   explicit[s=2]
   avg      1.5      0.5      0.5      0.8      0.9      0.7      1.1      0.5      1.1      0.0       7.6       5.4+/-  3.3     201.4   explicit[s=4]
   avg      2.0      0.6      0.5      0.9      1.0      0.7      1.3      0.6      1.2      0.0       8.7       6.0+/-  5.8     208.8   explicit[s=6]
   avg      2.6      0.6      0.5      1.0      1.1      0.8      1.8      0.6      1.3      0.0      10.4       7.0+/-  9.2     217.8   explicit[s=8]
```

This is compiled with `-d` and running it with `-w 1`.
```
master
Total Timing Summary
Average time for map loading = 593.076 ms
Average time for input loading = 7828.69 ms
Average time for lst::Event creation = 0.000669842 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     15.2      0.5      0.4      0.8      0.8      0.7      1.0      0.5      1.1      0.1      21.1       5.2+/-  0.9    2769.5   explicit_cutvalue[s=1]
   avg      1.2      0.5      0.4      0.8      0.8      0.7      0.9      0.5      1.1      0.1       6.9       5.1+/-  1.2    2596.9   explicit_cutvalue[s=2]
   avg      1.6      0.5      0.5      0.8      0.9      0.7      1.1      0.5      1.1      0.1       7.7       5.4+/-  3.0    2608.1   explicit_cutvalue[s=4]
   avg      2.2      0.6      0.5      0.9      1.0      0.7      1.1      0.7      1.2      0.1       8.8       6.0+/-  5.2    2770.6   explicit_cutvalue[s=6]
   avg      2.6      0.6      0.5      0.9      1.1      0.7      1.5      0.7      1.3      0.1      10.1       6.7+/-  7.9    2895.8   explicit_cutvalue[s=8]

This PR
Total Timing Summary
Average time for map loading = 595.089 ms
Average time for input loading = 7812.24 ms
Average time for lst::Event creation = 0.000669843 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     14.5      0.5      0.4      0.7      0.8      0.7      0.9      0.4      1.1      0.1      20.2       5.0+/-  0.9    4725.2   explicit_cutvalue[s=1]
   avg      1.3      0.5      0.4      0.8      0.8      0.7      0.9      0.5      1.1      0.1       7.1       5.1+/-  1.4    4952.4   explicit_cutvalue[s=2]
   avg      1.7      0.5      0.5      0.9      0.9      0.7      1.1      0.5      1.1      0.1       7.9       5.6+/-  3.1    4911.3   explicit_cutvalue[s=4]
   avg      2.1      0.6      0.5      0.9      1.0      0.7      1.3      0.6      1.2      0.1       9.0       6.2+/-  5.9    4970.4   explicit_cutvalue[s=6]
   avg      3.8      0.7      0.5      1.0      1.1      0.8      1.8      0.8      1.3      0.1      11.9       7.4+/-  9.1    4982.9   explicit_cutvalue[s=8]
```

slava77 commented:
> This is compiled with `-d` and running it with `-w 1`.

isn't the GNN part from `lst -G` and without recompilation with `-d`?

ariostas commented:
Also, @GNiendorf I'm getting random segfaults with `-G` even on `master`, so we'll have to look into that.

GNiendorf commented:
> > This is compiled with `-d` and running it with `-w 1`.
> 
> isn't the GNN part from `lst -G` and without recompilation with `-d`?

The NN training datasets are generated with -d and -w 1, I don't think anyone has used the -G option in years. See comment above, I suggest we remove it considering the changes in this PR.

alexandertuna commented:
Hi @ariostas @GNiendorf , I agree it's a good idea to test the NN training.

I'm looking at folding [this other PR](https://github.com/SegmentLinking/cmssw/pull/188/) into this PR and I see there are a handful of pLS branches which exist in master and don't exist here (pLS_isQuad, pLS_etaErr, pLS_circleRadius for example)

GNiendorf commented on RecoTracker/LSTCore/interface/SegmentsSoA.h:
```diff
@@ -15,6 +15,15 @@ namespace lst {
                       SOA_COLUMN(FPX, dPhiChanges),
                       SOA_COLUMN(FPX, dPhiChangeMins),
                       SOA_COLUMN(FPX, dPhiChangeMaxs),
+#ifdef CUT_VALUE_DEBUG
```
I forget how these flags are coded, but is there any way to make these separate compile-time flags, like SG_CUT_DEBUG, MD_CUT_DEBUG, etc. that are all triggered by the same -d option? This way people could disable flags for different objects if they aren't needed to help mitigate timing increases.

ariostas commented:
@aashayarora have you been using the GNN branches or is it okay if we remove them?

aashayarora commented:
> @aashayarora have you been using the GNN branches or is it okay if we remove them?

I have been using them, but I can create a new branch if you want to remove them.

ariostas commented:
I rebased to fix merge conflicts again. I also added more flags for more granular control. Each object/dnn has it's own flag. So you can do for example `lst_cuda --t3 --pls --t5dnn ...`. If you want everything you can use `--allobj`. The `-d` compilation flag still enables a few extra things. I left them as a compilation flag since it enlarges the SoAs, so it increases the vram requirement.

@GNiendorf I added the missing branches back. There were some that were already there, but were renamed.
t5_matched_simIdx -> t5_simIdxAll
pLS_matched_simIdx -> pLS_simIdxAll
pLS_ptIn -> pLS_pt
pT3_matched_simIdx -> pT3_simIdxAll

Also, I noticed that pT3_score was not even being filled in the original code.

When you have a chance, could you take a look to see if it looks good?

@kk428 I modified you jets code a bit, so could you check if I didn't break it?

GNiendorf commented:
Ahhh kk, thank you. I will check the DNN notebooks.

kk428 commented:
> @kk428 I modified you jets code a bit, so could you check if I didn't break it?

I checked the code, it still works!

ariostas commented:
Thank you, Gavin! I'll check if some branch is not being filled by mistake

GNiendorf commented:
https://gist.github.com/GNiendorf/d14dc8ee7778d3fc8e286eb74fd7ad98
https://gist.github.com/GNiendorf/5cd2bd1dcb53129253ff97ee3fdffaf7

Somewhat convoluted script that checks where there are differences in the branches between master ntuple and this PR ntuple. Seems like there are shape differences between the simidx branches for example, and some other differences that should be checked.

GNiendorf commented:
I guess for the pLS branches they are added differently than how I added them to the ntuple, so the order is not going to be the same.

GNiendorf commented:
Looks good, thanks @ariostas!

slava77 commented:
is the GPU more costly now at high stream count or is there some stability issue on the CI side?
<img width="152" height="214" alt="image" src="https://github.com/user-attachments/assets/ea1b6d8e-49e2-4773-a09e-3f8ef52a7fc0" />

(bottom is with this PR)

GNiendorf commented:
Timing looks good on lnx4555
<img width="1357" height="181" alt="Screenshot 2025-08-27 at 7 48 47 PM" src="https://github.com/user-attachments/assets/7f330dc3-6a37-48a8-bcb2-cbdd0981ad10" />

ariostas commented:
I'll have to look into this further because it seems to get stuck on event 129 when all outputs are enabled.

GNiendorf commented:
Where is the timing increase coming from? Are we storing more info into the standard ntuple, or is the code just slower? If it is the former, don't we have an -l flag that toggles saving low-level info in the ntuple? We could put the new branches for the low-level objects under that flag so we don't save them by default if so.

ariostas commented:
The output ntuple is still more than 4 times larger, so that must be it. I'll see which branches could be put behind flags so that it doesn't save so much stuff by default.

GNiendorf commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -313,296 +527,1261 @@ void setOutputBranches(LSTEvent* event) {
+        }
+      }
+    }
+    pT3_isDuplicate[i] = isDuplicate;
```
Is this also a mistake? In the t3 branches but setting pT3_isDuplicate, and same above. `std::vector<int> pT3_isDuplicate(t3_simIdxAll.size());`

ariostas replied:
Oh yeah, I don't know how that extra vector got there

alexandertuna commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -313,296 +527,1261 @@ void setOutputBranches(LSTEvent* event) {
+        int hitidx = trk_see_hitIdx[ipLS][ihit];
+        int hittype = trk_see_hitType[ipLS][ihit];
+        int x = trk_pix_x[hitidx];
+        int y = trk_pix_x[hitidx];
```
😬  That's not good @ariostas is it possible to check ALL the branches with ntuple_checker.py ?

ariostas replied:
Good catch! @alexandertuna we can't check the new branches since there's nothing to compare to.

GNiendorf commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -313,296 +527,1261 @@ void setOutputBranches(LSTEvent* event) {
+  // printT3s(event);
+  for (unsigned int idx = modules.nLowerModules(); idx <= modules.nLowerModules(); ++idx) {
+    unsigned int nmods = modules.nLowerModules();
+    for (unsigned int ipLS = 0; ipLS < segmentsOccupancy.nSegments()[idx]; ipLS++) {
```
Tried to make a comment here not sure where it went, but I don't understand how this loop can give the same results as the master? We iterate over the lower modules and push a lot of features using ipLS as the index when accessing the arrays, but ipLS doesn't depend on the lower module?

GNiendorf replied:
Maybe we just revert this 2d loop structure to the 1d one on master?

ariostas replied:
Yeah, this is pretty horrible. The only reason why this works is because the outer loop only goes through one iteration, so there wasn't even a point in having it.

There's some other non-ideal things I've seen, which is why I mentioned doing some further cleanup later on.

GNiendorf replied:
Oh yeah I just saw the outer loop starts at nLowerModules haha. kk.

ariostas commented on RecoTracker/LSTCore/standalone/code/core/write_lst_ntuple.cc:
```diff
@@ -76,112 +24,73 @@ void createOptionalOutputBranches() {
+  if (ana.t3dnn_branches)
+    setT3DNNBranches(event, matchfrac);
+  if (ana.t5_branches)
+    setT5DNNBranches(event);
```
Oh yeah, that was a typo

aashayarora commented:
> @aashayarora were you running with `--allobj` or just with the standard outputs?

I was running with the `--md` and `--ls` flags. 
The exact command I ran is `lst -i PU200RelVal -s 32 -o output_pu200.root --md --ls`

ariostas commented:
Okay, hopefully everything is good now.

I fixed the issues Gavin pointed out, moved some extra branches behind flags, and fixed the memory leak.

Here's the new timing for the standard outputs on cgpu-1. The ntuple is now only about 10% larger.
```
This PR
Total Timing Summary
Average time for map loading = 582.945 ms
Average time for input loading = 18068.4 ms
Average time for lst::Event creation = 0.000579016 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     16.7      0.4      0.5      1.0      1.1      0.7      0.9      0.5      1.1      0.0      23.0       5.6+/-  1.6      23.0   explicit[s=1]
   avg      2.6      0.9      0.8      1.4      1.4      0.7      1.5      0.8      1.8      0.0      12.0       8.6+/-  2.1       6.0   explicit[s=2]
   avg      3.0      1.7      1.9      2.7      2.5      1.0      2.6      1.1      3.1      0.0      19.6      15.6+/-  2.8       5.0   explicit[s=4]
   avg      4.1      2.4      3.0      3.8      4.1      1.5      4.0      1.7      4.6      0.0      29.2      23.7+/-  5.0       5.0   explicit[s=6]
   avg      5.2      3.5      4.2      5.5      5.5      1.7      5.1      2.4      5.8      0.0      38.9      32.0+/-  6.8       5.0   explicit[s=8]

master
Total Timing Summary
Average time for map loading = 579.196 ms
Average time for input loading = 7751.67 ms
Average time for lst::Event creation = 0.00055631 ms
   Evt    Hits       MD       LS      T3       T5       pLS       pT5      pT3      TC       Reset    Event     Short             Rate
   avg     15.9      0.5      0.5      1.0      1.1      0.7      0.9      0.5      1.1      0.0      22.1       5.6+/-  1.6      22.1   explicit[s=1]
   avg      2.6      0.9      0.9      1.4      1.5      0.7      1.5      0.8      1.7      0.0      12.0       8.7+/-  2.0       6.1   explicit[s=2]
   avg      2.9      1.6      1.9      2.6      2.6      1.0      2.6      1.2      3.1      0.0      19.5      15.6+/-  3.1       5.0   explicit[s=4]
   avg      4.0      2.6      2.9      3.9      4.1      1.4      4.0      1.7      4.6      0.0      29.2      23.8+/-  5.1       5.0   explicit[s=6]
   avg      5.1      3.4      4.1      5.5      5.6      1.7      4.7      2.4      5.8      0.0      38.3      31.6+/-  4.4       4.9   explicit[s=8]
```

/run gpu-all
